### PR TITLE
Sampling

### PR DESCRIPTION
**Pull request by jungikim**
Implements sampled dataset.
Sampling can be done via uniform or perplexity-based multinomial probability distribution.

perplexity-based sampling is experimental (effects to be experimented).

When perplexity-based sampling is required, an optional return value indvLoss is passed up from  decoder -> Seq2Seq -> trainer.

### CODE REVIEW

**Comment by jungikim** [comment]
As-is, this PR introduces an additional dependency on torchx.
We can avoid it if necessary by only including the necessary module from https://github.com/nicholas-leonard/torchx/blob/master/AliasMultinomial.lua (BSD-3-Clause)

File: onmt/modules/Decoder.lua [hunk 1]
```diff
@@ -329,14 +330,25 @@ function Decoder:backward(batch, outputs, criterion)
...
     -- Compute decoder output gradients.
     -- Note: This would typically be in the forward pass.
     local pred = self.generator:forward(outputs[t])
     local output = batch:getTargetOutput(t)
 
-    loss = loss + criterion:forward(pred, output)
+    if self.args.indvLoss then
+      local curLossSum = 0
+      for i=1, pred[1]:size(1) do
+        local tmpLoss = criterion:forward({pred[1][i]},{output[1][i]})
```
Comment (guillaumekln) [inline onmt/modules/Decoder.lua]: This does not work with target features. For example with one target feature:

`pred` is in the form:

```
{
  1 : CudaTensor - size: 64x34058
  2 : CudaTensor - size: 64x9
}
```

and `output`:

```
{
  1 : CudaLongTensor - size: 64
  2 : CudaLongTensor - size: 64
}
```

File: onmt/modules/Decoder.lua [hunk 2]
```diff
@@ -365,8 +377,12 @@ function Decoder:backward(batch, outputs, criterion)
       gradStatesInput[i]:copy(gradInput[i])
     end
   end
+  
+  if self.args.indvLoss then
+    indvAvgLoss:div(batch.targetLength)
```
Comment (guillaumekln) [inline onmt/modules/Decoder.lua]: Shouldn't each loss be normalized by each target true length?

File: onmt/data/SampledDataset.lua [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+      if self.ppl[i] ~= self.sample_w_ppl_init and self.ppl[i] > threshold then
+        -- asign low value to instances with ppl above threshold (outliers)
+        self.samplingProb[i] = 1
+      else
+        self.samplingProb[i] = self.ppl[i]
+      end
+    end
+  end
+
+--  local sampled = torch.multinomial(self.samplingProb:view(1,#self.src), self.samplingSize, --[[replacement]] false)
```
Comment (jungikim) [inline onmt/data/SampledDataset.lua]: I'd like to keep this comment even if it is dead code just as a reminder of how it is done by generic torch library. Maybe the torch function will be updated for speed improvement in the future.

File: onmt/data/SampledDataset.lua [hunk 4]
```diff
@@ -0,0 +1,266 @@
+--[[ Data management and batch creation. Handles data created by `preprocess.lua`. ]]
+require 'torchx'
+
+local SampledDataset = torch.class("SampledDataset")
```
Comment (jungikim) [inline onmt/data/SampledDataset.lua]: With changes made in e276661, the batch prep logic diverged more and I don't think there's a room for sharing.

**Comment by jungikim** [comment]
> Did you test for efficiency? I measured a 20% slowdown when using -sample_w_ppl with the default options. I guess that comes from the criterion forward step. Not a big deal but this should be taken into account as this dev aims to make the overall training faster.

In my 1mil train, ppl-based sampling is ~10% slower than random sampling due to, as you mentioned, criterion forward step. As stated in the PR request, ppl-based sampling is experimental and the speed up is expected in the reduced number of iterations or from enabling the model better handle more difficult training instances.

File: onmt/train/Trainer.lua [hunk 5]
```diff
@@ -254,10 +271,24 @@ function Trainer:train(model, optim, trainData, validData, dataset, info)
       end
     end
 
+    -- if needed, get avgPpl from epochState before log() since log() will reset internal stats
+    local avgPpl
+    if self.options.sample and self.options.sample_w_ppl then
+      avgPpl = epochState:getAvgPpl()
```
Comment (jungikim) [inline onmt/train/Trainer.lua]: Reverted with 8c3fcf3. I didn't know epochState keeps only the stat from last batch, The module name implies otherwise.

File: onmt/modules/Decoder.lua (hunk 1 above)
Comment (jungikim) [inline onmt/modules/Decoder.lua]: fixed in 9630b83

File: onmt/modules/Decoder.lua (hunk 2 above)
Comment (jungikim) [inline onmt/modules/Decoder.lua]: fixed in 9630b83

File: onmt/modules/Decoder.lua [hunk 6]
```diff
@@ -41,6 +41,7 @@ function Decoder:__init(inputNetwork, rnn, generator, inputFeed)
   -- vector each time representing the attention at the
   -- previous step.
   self.args.inputFeed = inputFeed
+  self.args.indvLoss = indvLoss
```
Comment (jungikim) [inline onmt/modules/Decoder.lua]: Good idea not adding it in serialized attribute. Fixed in 4c70ca2.

**Comment by jungikim** [comment]
@guillaumekln Updated branch passes luacheck, test/test.lua, and runs on sample data with/without case feature.
I think this is ready for another review.

File: train.lua [hunk 7]
```diff
@@ -11,6 +11,13 @@ local modelClass = onmt.ModelSelector(modelType)
 local options = {
   {'-data',       '', [[Path to the training *-train.t7 file from preprocess.lua]],
                       {valid=onmt.utils.ExtendedCmdLine.nonEmpty}},
+
+  {'-sample',              0, [[Number of instances to sample from train data in each epoch]]},
+  {'-sample_w_ppl',    false, [[use ppl as probability distribution when sampling]]},
+  {'-sample_w_ppl_init',  15, [[start perplexity-based sampling when average train perplexity per batch falls below this value]]},
+  {'-sample_w_ppl_max', -1.5, [[when greater than 0, instances with perplexity above this value will be considered as noise and ignored;
+                                when less than 0, mode + (-sample_w_ppl_max) * stdev will be used as threshold]]},
+
```
Comment (jsenellart-systran) [inline train.lua]: these options should be moved, probably to SampleDataSet

File: onmt/data/SampledDataset.lua [hunk 8]
```diff
@@ -0,0 +1,266 @@
+--[[ Data management and batch creation. Handles data created by `preprocess.lua`. ]]
+require 'torchx'
```
Comment (jsenellart-systran) [inline onmt/data/SampledDataset.lua]: considering the size of AliasMultinomial.lua - we should rather integrate it in onmt as we used to integrate LinearNoBias - waiting for the module to pass into standard libraries

File: onmt/data/SampledDataset.lua [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+local SampledDataset = torch.class("SampledDataset")
+
+--[[ Initialize a data object given aligned tables of IntTensors `srcData`
+  and `tgtData`.
+--]]
+function SampledDataset:__init(srcData, tgtData, samplingSize, sample_w_ppl, sample_w_ppl_init, sample_w_ppl_max)
+
+  self.src = srcData.words
```
Comment (jsenellart-systran) [inline onmt/data/SampledDataset.lua]: @Jungi - in #142, I introduce the ability to replace word input by direct feature input (srcData.vectors) - here your code fully assume that it is words, can you anticipate this change and make your code more generic?

File: onmt/Seq2Seq.lua [hunk 10]
```diff
@@ -73,6 +73,16 @@ function Seq2Seq.dataType()
   return 'bitext'
 end
 
+function Seq2Seq:returnIndividualLosses(enable)
+  if torch.getmetatable(torch.type(self.models.decoder))['returnIndividualLosses'] == nil then
```
Comment (guillaumekln) [inline onmt/Seq2Seq.lua]: This could be simplified to:

```lua
if not self.models.decoder.returnIndividualLosses then
```

File: train.lua [hunk 11]
```diff
@@ -107,6 +128,13 @@ local function main()
     end
   end)
 
+  if opt.sample_w_ppl then
+    if torch.getmetatable(torch.type(model))['returnIndividualLosses'] == nil or model:returnIndividualLosses(true) == false then
```
Comment (guillaumekln) [inline train.lua]: Same simplification as above:

```lua
if not model.returnIndividualLosses
```

File: onmt/data/SampledDataset.lua (hunk 3 above)
Comment (guillaumekln) [inline onmt/data/SampledDataset.lua]: OK.

File: onmt/data/SampledDataset.lua (hunk 4 above)
Comment (guillaumekln) [inline onmt/data/SampledDataset.lua]: OK.

File: onmt/data/SampledDataset.lua (hunk 8 above)
Comment (guillaumekln) [inline onmt/data/SampledDataset.lua]: OK. Let's put it in the `onmt/data` directory then with the license in the file header.

File: onmt/Seq2Seq.lua (hunk 10 above)
Comment (jungikim) [inline onmt/Seq2Seq.lua]: Fixed in 08ad2ed

File: train.lua (hunk 11 above)
Comment (jungikim) [inline train.lua]: Fixed in 08ad2ed

File: train.lua (hunk 7 above)
Comment (jungikim) [inline train.lua]: I moved the options to SampledDataset as suggested (08ad2ed)

File: onmt/data/SampledDataset.lua (hunk 9 above)
Comment (jungikim) [inline onmt/data/SampledDataset.lua]: Retrieving main feature from srcData and tgtData is now changed to "srcData.words or srcData.vectors" and  "tgtData.words or tgtData.vectors"; I hope this is enough. (08ad2ed)

File: onmt/data/SampledDataset.lua (hunk 3 above)
Comment (jungikim) [inline onmt/data/SampledDataset.lua]: Thank you.

File: onmt/data/SampledDataset.lua (hunk 4 above)
Comment (jungikim) [inline onmt/data/SampledDataset.lua]: I reverted back to preparing batch by iterating over all instances due to batch efficiency and to keep correct attention weights without using maskedsoftmax.

However, to allow oversampling, Instead of adding one instance when it is marked as sampled, now instances can be added multiple time.

This will potentially bias the batch towards more frequently sampled ones and will have more effect (whether positive or negative) on training than before.

File: onmt/data/SampledDataset.lua (hunk 8 above)
Comment (jungikim) [inline onmt/data/SampledDataset.lua]: Added: 186a2ad

**Comment by guillaumekln** [comment]
You should update the tests also.

**Comment by jungikim** [comment]
I fixed test (6d5da5e)
